### PR TITLE
cmake: improve package config file

### PR DESCRIPTION
**Pull request by ygj6**
After this patch, one can use `find_package()` to integrate libevent (both build tree and install tree) into their own project.
usage:
```
set(LIBEVENT_STATIC_LINK TRUE)
find_package(Libevent 2.2 REQUIRED COMPONENTS event_core event_extra)
include_directories(${LIBEVENT_INCLUDE_DIRS})
target_link_libraries(myapp ${LIBEVENT_LIBRARIES})
```

Part of idea comes from this [code](https://github.com/psi4/gdma/blob/master/cmake/gdmaConfig.cmake.in)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1470,23 +1463,13 @@ set(LIBEVENT_INCLUDE_DIRS
     ${EVENT__INCLUDE_DIRS}
     CACHE PATH "Libevent include directories")
 
-configure_file(${PROJECT_SOURCE_DIR}/cmake/LibeventConfigBuildTree.cmake.in
+set(CONFIG_FOR_INSTALL_TREE 0)
+configure_file(${PROJECT_SOURCE_DIR}/cmake/LibeventConfig.cmake.in
                ${PROJECT_BINARY_DIR}/LibeventConfig.cmake
                @ONLY)
```
Comment (azat) [inline CMakeLists.txt]: Wrapping this into function would be nice

File: CMakeLists.txt [hunk 2]
```diff
@@ -1449,15 +1449,8 @@ endif()
...
-if(WIN32 AND NOT CYGWIN)
-  set(DEF_INSTALL_CMAKE_DIR cmake)
-else()
-  set(DEF_INSTALL_CMAKE_DIR lib/cmake/libevent)
-endif()
-
 set(EVENT_INSTALL_CMAKE_DIR
-    "${CMAKE_INSTALL_PREFIX}/${DEF_INSTALL_CMAKE_DIR}"
-    CACHE PATH "Installation directory for CMake files")
+    "${CMAKE_INSTALL_PREFIX}/lib/cmake/libevent")
```
Comment (azat) [inline CMakeLists.txt]: So we cannot change this from the cmake args - ok

File: cmake/LibeventConfig.cmake.in [hunk 3]
```diff
@@ -1,18 +1,178 @@
...
 #  LIBEVENT_STATIC_LIBRARIES - libraries to link against (archive/static)
 #  LIBEVENT_SHARED_LIBRARIES - libraries to link against (shared)
+#
+# When you try to locate the libevent libraries, you should specify which components you want to use.
+# The following table lists all available components. If none is given, all imported targets will used.
+#  event_core        - the core functons of libevent
+#  event_extra       - extra functions, contains http, dns ans rpc
+#  event             - sum of two above, deprecated
+#  event_pthreads    - multiple threads for libevent, not exists on Windows
+#  event_openssl     - openssl support for libevent
```
Comment (azat) [inline cmake/LibeventConfig.cmake.in]: Hm, don't you think stripping `event_` from the name would be better?

File: cmake/LibeventConfig.cmake.in [hunk 4]
```diff
@@ -1,18 +1,178 @@
...
+#
+# These variables are deprecated, only reserved for compatibility.
 #  LIBEVENT_STATIC_LIBRARIES - libraries to link against (archive/static)
 #  LIBEVENT_SHARED_LIBRARIES - libraries to link against (shared)
+#
+# When you try to locate the libevent libraries, you should specify which components you want to use.
+# The following table lists all available components. If none is given, all imported targets will used.
+#  event_core        - the core functons of libevent
+#  event_extra       - extra functions, contains http, dns ans rpc
+#  event             - sum of two above, deprecated
```
Comment (azat) [inline cmake/LibeventConfig.cmake.in]: I think that should not be here (to not encourage using of deprecates)

File: cmake/LibeventConfig.cmake.in (hunk 3 above)
Comment (azat) [reply cmake/LibeventConfig.cmake.in]: Like `boost`

File: cmake/LibeventConfig.cmake.in [hunk 5]
```diff
@@ -1,18 +1,178 @@
...
+    else()
+        message_if_needed(WARNING "Can not find any libraries for libevent.")
+    endif()
+endif()
+
+unset(_LIB_TYPE)
+unset(_AVAILABLE_LIBS)
+unset(_EVENT_COMPONENTS)
+unset(_POSSIBLE_PKG_NAMES)
+unset(_INSTALL_PREFIX)
```
Comment (azat) [inline cmake/LibeventConfig.cmake.in]: It will be nice to have a test for this on our CI (with build-tree and install-tree and system-wide install)

File: cmake/LibeventConfig.cmake.in [hunk 6]
```diff
@@ -1,18 +1,178 @@
...
+#  event_extra       - extra functions, contains http, dns ans rpc
+#  event             - sum of two above, deprecated
+#  event_pthreads    - multiple threads for libevent, not exists on Windows
+#  event_openssl     - openssl support for libevent
+#
+# By default, the shared libraries of libevent will be found. To find the static ones instead,
+# you must set the LIBEVENT_STATIC_LINK variable to TRUE before calling find_package(libevent ...).
+# example:
+#  set(LIBEVENT_STATIC_LINK TRUE)
+#  find_package(libevent 2.2 REQUIRED COMPONENTS event_core event_extra)
```
Comment (azat) [inline cmake/LibeventConfig.cmake.in]: Just a note, it will be nice to export `PACKAGE_VERSION` with the number of commits since the last tag, to make the following possible:

```
find_package(libevent 2.2.xx REQUIRED) 
```

(But should be a separate patch)

File: CMakeLists.txt (hunk 1 above)
Comment (ygj6) [reply CMakeLists.txt]: Extracted a macro `gen_package_config()`

File: CMakeLists.txt (hunk 2 above)
Comment (ygj6) [reply CMakeLists.txt]: Consider these steps:
```
cmake -DCMAKE_INSTALL_PREFIX=/first/install/path ..
cmake -DCMAKE_INSTALL_PREFIX=/second/install/path ..
cmake --build . --target install
```
The final products are always installed to /first/install/path because `EVENT_INSTALL_CMAKE_DIR` was cached. So I removed the `Cache` key.

File: cmake/LibeventConfig.cmake.in (hunk 3 above)
Comment (ygj6) [reply cmake/LibeventConfig.cmake.in]: The new usage:
```
set(LIBEVENT_STATIC_LINK TRUE)
find_package(libevent 2.2 REQUIRED COMPONENTS core extra openssl pthreads)
target_link_libraries(myapp ${LIBEVENT_LIBRARIES})
```

Comment (ygj6) [reply cmake/LibeventConfig.cmake.in]: After stripping `event_ `, the variables `LIBEVENT_STATIC_LIBRARIES` and `LIBEVENT_SHARED_LIBRARIES` are not functional now.

File: cmake/LibeventConfig.cmake.in (hunk 4 above)
Comment (ygj6) [reply cmake/LibeventConfig.cmake.in]: We no longer export `event`. And one can not use `find_package()` to link it.

File: cmake/LibeventConfig.cmake.in (hunk 6 above)
Comment (ygj6) [reply cmake/LibeventConfig.cmake.in]: I'll look at this.

File: cmake/LibeventConfig.cmake.in (hunk 5 above)
Comment (ygj6) [reply cmake/LibeventConfig.cmake.in]: This function is difference from others. For testing, should I install the products and create a standalone project?

Comment (azat) [reply cmake/LibeventConfig.cmake.in]: >should I install the products

All available combinations should be covered:
- install into system-wide path
- install into non-system-wide path
- build tree

> and create a standalone project?

I would say that this should be in libevent tree, since otherwise it will require extra work to make it run for each commit (i.e. `sample/libevent-export-test` or something similar, also this will give an example for users)

File: CMakeLists.txt [hunk 7]
```diff
@@ -1449,18 +1451,24 @@ endif()
...
 
 export(PACKAGE libevent)
 
+macro(gen_package_config forinstall)
```
Comment (azat) [inline CMakeLists.txt]: AFAICS function can be used (to reduce variables scope)

**Comment by azat** [comment]
Not finished yet (ci build canceled)?

File: test-export/test-export.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
+#
+# Check if find_package(Libevent COMPONENTS xxx) can get the correct library.
+# Note: this script has only been tested on python3.
+# Usage:
+#   cd cmake-build-dir
+#   cmake <options> .. && cmake --build .
+#   python /path/to/test-export.py [static|shared]
```
Comment (azat) [inline test-export/test-export.py]: Wow, coverage is decent, any reasons for python been used over simple shell?

Comment (azat) [reply test-export/test-export.py]: And the reason calls "windows", got it

File: test-export/test-export.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    if r == 0:
+        r = exec_cmd('cmake --build .', True)
+        if r == 0:
+            r = exec_cmd('ctest', True)
+    if r != 0:
+        r = 1
+    return r
+
+# expect  0:success 1:failure
+def testcase(link, code, expect):
```
Comment (azat) [inline test-export/test-export.py]: if written in python, can be done (totally up to you to decide, since the profit from using it here is logging and parametrization, while later is an extra dependency) using unittest framework

Comment (ygj6) [reply test-export/test-export.py]: I have tried to implement with unittest framework: https://github.com/ygj6/libevent/commit/0fba83826be8435a005de8a1fad796edbd4be8be
However, [appveyor](https://ci.appveyor.com/project/ygj6/libevent/builds/28884314/job/edu72ucbh051wl3x) always fails because it treats the outputs of unittest as errors.
So I have to fall back to the previous.
